bin/mysql.sh

```diff
@@ -0,0 +1,386 @@
+        if [[ ${PRIMARY_HOST} != "null" ]] && [[ -n ${PRIMARY_HOST} ]]; then
+            echo 'Found primary at:' ${PRIMARY_HOST}
+            break
+        fi
+        echo -ne '.'
+        sleep 1
+    done
+}
+
+# returns "true" on success or "false" on failure due to CAS
+markSelfPrimary() {
+    PRIMARY_NODE=mysql-$(hostname)
+    echo "Marking ${PRIMARY_NODE} as primary..."
+    for i in {30..0}; do
+        local ok=$(curl -XPUT -s --fail -d ${PRIMARY_NODE} \
+                        "http://${CONSUL}:8500/v1/kv/mysql-primary?cas=0" | jq -r '.')
+        if [[ ${ok} = "true" ]]; then
+            echo
+            break
+        fi
+        if [[ ${ok} = "false" ]]; then
+            # The only way this scenario appears is if we raced with
+            # another node. So we're going to just start over.
+            echo 'Tried to set this instance as primary but primary exists.'
+            echo 'Restarting bootstrap process.'
+            onStart
+        fi
+        echo -ne '.'
+        sleep 1
+    done
+}
+
+# Set up GITD-based replication to the primary; once this is set the
+# replica will automatically try to catch up with the primary by pulling
+# its entire binlog. This is comparatively slow but suitable for small
+# databases.
+setPrimaryForReplica() {
+    "${mysql[@]}" <<-EOSQL
+		CHANGE MASTER TO
+		MASTER_HOST           = '$PRIMARY_HOST',
+		MASTER_USER           = '$MYSQL_REPL_USER',
+		MASTER_PASSWORD       = '$MYSQL_REPL_PASSWORD',
+		MASTER_PORT           = 3306,
+		MASTER_CONNECT_RETRY  = 60,
+        MASTER_AUTO_POSITION  = 1,
+		MASTER_SSL            = 0;
+        START SLAVE;
+EOSQL
+
+}
+
+stopReplica() {
+    echo 'STOP SLAVE;' | "${mysql[@]}"
+}
+
+replica() {
+    checkConfig        # make sure we have all required environment variables
+    initializeDb       # sets datadir
+    bootstrapDb        # $PID for temporary mysqld while we bootstrap
+    waitForConnection  # sets $mysql[] which we'll use for all future conns
+
+    mysql_tzinfo_to_sql /usr/share/zoneinfo | "${mysql[@]}" mysql
+
+    echo 'Setting up replication.'
+
+    # belt-and-suspenders in the workflow so that if we call this
+    # script with `replica` as the command we assert environment
+    if [[ -z ${PRIMARY_HOST} ]]; then
+        getPrimaryHost
+        if [[ -z ${PRIMARY_HOST} ]]; then
+            echo 'Tried replication setup, but primary is set and not healthy. Exiting!'
+            exit 1
+        fi
+    fi
+
+    setPrimaryForReplica
+    cleanup
+}
+
+
+# TODO: this section will be for cases where we have a large existing
+# dataset that we need to update from the master before starting
+# replication. This way we avoid running thru the entire binlog, but
+# it requires outside intervention that we won't be able to do easily
+# just via docker-compose
+
+# dataDump() {
+#     mysqldump -h ${PRIMARY_HOST} -p 3306 --all-databases --master-data --set-gtid-purged=ON > dbdump.db
```

Comment:
This may just be me, but my preferred way to do this is to stop MySQL and grab the filesystem. My reasoning for that includes:

1. The only safe way to get a dump is to lock the tables, which is effectively the same as stopping MySQL.
1. Dumps and imports take a _very_ long time compared to filesystem copies.